velox/core/Expressions.cpp

```diff
@@ -465,7 +466,7 @@ TypedExprPtr FieldAccessTypedExpr::rewriteInputNames(
   VELOX_CHECK_EQ(1, newInputs.size());
   // Only rewrite name if input in InputTypedExpr. Rewrite in other
   // cases(like dereference) is unsound.
-  if (!std::dynamic_pointer_cast<const InputTypedExpr>(newInputs[0])) {
+  if (!is_instance_of<InputTypedExpr>(newInputs[0])) {
```

Comment:
@juwentus1234 @xiaoxmeng 

As discussed earlier, we should not use cast here. Instead, use `newInputs[0]->isInputKind()`. Ditto other places.